File: sdk/python/bunqueue/simple/cancellation.py
"""Graceful job cancellation (mirror of cancellation.ts).

Python has no AbortController; :class:`CancelSignal` exposes the same
``aborted`` surface, checked cooperatively by processors and the middleware
chain.
"""

from __future__ import annotations

import threading
from typing import Dict, Optional


class CancelSignal:
    """Minimal AbortSignal equivalent: ``signal.aborted`` flips once."""

    def __init__(self) -> None:
        self._event = threading.Event()

    @property
    def aborted(self) -> bool:
        return self._event.is_set()

    def abort(self) -> None:
        self._event.set()


class CancellationManager:
    def __init__(self) -> None:
        self._signals: Dict[str, CancelSignal] = {}
        self._timers: Dict[str, threading.Timer] = {}
        self._lock = threading.Lock()

    def register(self, job_id: str) -> CancelSignal:
        signal = CancelSignal()
        with self._lock:
            self._signals[job_id] = signal
        return signal

    def unregister(self, job_id: str) -> None:
        with self._lock:
            self._signals.pop(job_id, None)
            timer = self._timers.pop(job_id, None)
        if timer:
            timer.cancel()

    def cancel(self, job_id: str, grace_period_ms: float = 0) -> None:
        with self._lock:
            signal = self._signals.get(job_id)
        if signal is None:
            return
        if grace_period_ms > 0:
            timer = threading.Timer(grace_period_ms / 1000.0, signal.abort)
            timer.daemon = True
            with self._lock:
                self._timers[job_id] = timer
            timer.start()
        else:
            signal.abort()

    def is_cancelled(self, job_id: str) -> bool:
        with self._lock:
            signal = self._signals.get(job_id)
        return signal.aborted if signal else False

    def get_signal(self, job_id: str) -> Optional[CancelSignal]:
        with self._lock:
            return self._signals.get(job_id)

    def destroy_all(self) -> None:
        with self._lock:
            signals = list(self._signals.values())
            timers = list(self._timers.values())
            self._signals.clear()
            self._timers.clear()
        for timer in timers:
            timer.cancel()
        for signal in signals:
            signal.abort()

File: src/client/bunqueue/cancellation.ts
/**
 * Bunqueue — Graceful Job Cancellation
 * AbortController-based cancellation with optional grace period.
 */

interface CancellationRegistration {
  readonly jobId: string;
  readonly controller: AbortController;
  timer: ReturnType<typeof setTimeout> | null;
  deadline: number | null;
}

export class CancellationManager {
  private readonly currentByJob = new Map<string, CancellationRegistration>();
  private readonly registrations = new Map<AbortController, CancellationRegistration>();

  /** Register a new AbortController for a job */
  register(jobId: string): AbortController {
    const controller = new AbortController();
    const registration = { jobId, controller, timer: null, deadline: null };
    this.currentByJob.set(jobId, registration);
    this.registrations.set(controller, registration);
    return controller;
  }

  /** Remove a job's controller (on completion) */
  unregister(jobId: string, controller?: AbortController): void {
    const registration = controller
      ? this.registrations.get(controller)
      : this.currentByJob.get(jobId);
    if (!registration || registration.jobId !== jobId) return;
    this.clearTimer(registration);
    this.registrations.delete(registration.controller);
    if (this.currentByJob.get(jobId) === registration) this.currentByJob.delete(jobId);
  }

  /** Cancel a job with optional grace period */
  cancel(jobId: string, gracePeriodMs = 0): void {
    const registration = this.currentByJob.get(jobId);
    if (!registration) return;

    if (gracePeriodMs <= 0) {
      this.clearTimer(registration);
      registration.controller.abort();
      return;
    }

    if (registration.controller.signal.aborted) return;
    const deadline = Date.now() + gracePeriodMs;
    if (registration.deadline !== null && registration.deadline <= deadline) return;
    this.clearTimer(registration);

    const timer = setTimeout(() => {
      if (registration.timer !== timer) return;
      registration.timer = null;
      registration.deadline = null;
      registration.controller.abort();
    }, gracePeriodMs);
    registration.timer = timer;
    registration.deadline = deadline;
  }

  /** Check if a job is cancelled */
  isCancelled(jobId: string): boolean {
    return this.currentByJob.get(jobId)?.controller.signal.aborted ?? false;
  }

  /** Get the AbortSignal for a job */
  getSignal(jobId: string): AbortSignal | null {
    return this.currentByJob.get(jobId)?.controller.signal ?? null;
  }

  /** Cancel all and clear */
  destroyAll(): void {
    const registrations = [...this.registrations.values()];
    this.currentByJob.clear();
    this.registrations.clear();
    for (const registration of registrations) this.clearTimer(registration);
    for (const registration of registrations) registration.controller.abort();
  }

  private clearTimer(registration: CancellationRegistration): void {
    const timer = registration.timer;
    registration.timer = null;
    registration.deadline = null;
    if (timer !== null) clearTimeout(timer);
  }
}
